tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java

```diff
@@ -254,64 +260,51 @@ public void endFile(Schema schema) throws IOException {
     // close file
     out.close();
     canWrite = false;
+    timeseriesMetadataMap = new TreeMap<>();
     logger.info("output stream is closed");
   }
 
   /**
-   * 1. group chunkGroupMetaDataList to TsDeviceMetadata 2. flush TsDeviceMetadata 3. get
-   * TsDeviceMetadataIndex
-   *
-   * @param chunkGroupMetaDataList all chunk group metadata in memory
-   * @return TsDeviceMetadataIndex in TsFileMetaData
+   * @return tsOffsets in TsFileMetaData
    */
-  private Map<String, TsDeviceMetadataIndex> flushTsDeviceMetaDataAndGetIndex(
-      List<ChunkGroupMetaData> chunkGroupMetaDataList) throws IOException {
-
-    Map<String, TsDeviceMetadataIndex> tsDeviceMetadataIndexMap = new HashMap<>();
-
-    long offset; /* offset for the flushing TsDeviceMetadata */
-
-    TsDeviceMetadata currentTsDeviceMetadata;
-
-    // flush TsDeviceMetadata by string order of deviceId
-    for (Map.Entry<String, TsDeviceMetadata> entry : getAllTsDeviceMetadata(chunkGroupMetaDataList)
-        .entrySet()) {
-      // update statistics in TsDeviceMetadata
-      currentTsDeviceMetadata = entry.getValue();
-
-      // flush tsChunkGroupBlockMetaData
-      offset = out.getPosition();
-      int size = currentTsDeviceMetadata.serializeTo(out.wrapAsStream());
-
-      TsDeviceMetadataIndex tsDeviceMetadataIndex = new TsDeviceMetadataIndex(offset, size,
-          currentTsDeviceMetadata);
-      tsDeviceMetadataIndexMap.put(entry.getKey(), tsDeviceMetadataIndex);
+  private long[] flushAllChunkMetadataList() throws IOException {
+    if (timeseriesMetadataMap.size() == 0) {
+      return new long[0];
     }
-
-    return tsDeviceMetadataIndexMap;
-  }
-
-  /**
-   * group all chunk group metadata by device.
-   *
-   * @param chunkGroupMetaDataList all chunk group metadata
-   * @return TsDeviceMetadata of all devices
-   */
-  private TreeMap<String, TsDeviceMetadata> getAllTsDeviceMetadata(
-      List<ChunkGroupMetaData> chunkGroupMetaDataList) {
-    String currentDevice;
-    TreeMap<String, TsDeviceMetadata> tsDeviceMetadataMap = new TreeMap<>();
-
-    for (ChunkGroupMetaData chunkGroupMetaData : chunkGroupMetaDataList) {
-      currentDevice = chunkGroupMetaData.getDeviceID();
-
-      if (!tsDeviceMetadataMap.containsKey(currentDevice)) {
-        TsDeviceMetadata tsDeviceMetadata = new TsDeviceMetadata();
-        tsDeviceMetadataMap.put(currentDevice, tsDeviceMetadata);
+    // convert ChunkMetadataList to this field
+    long[] tsOffsets = new long[timeseriesMetadataMap.size() + 1];
+    List<TimeseriesMetaData> tsMetadataList = new ArrayList<>();
+    // flush timeseriesMetadataList one by one
+    int i = 0;
+    for (Map.Entry<Path, List<ChunkMetaData>> entry : timeseriesMetadataMap.entrySet()) {
+      Path path = entry.getKey();
+      String deviceId = path.getDevice();
+      if (!deviceOffsetsMap.containsKey(deviceId)) {
+        deviceOffsetsMap.put(deviceId, new int[2]);
+        deviceOffsetsMap.get(deviceId)[0] = i;
+      }
```

Comment:
Please use the following codes instead to reduce the operation and improve the efficiency.
```
deviceOffsetsMap.computeIfAbsent(deviceId, k -> new int[2])[0] = i;
```

Reply:
This map is also changed.